File: DataStructures/Queue/Queue.java
public class Queue {
    private int[] arr;
    private int inicio, fim, size, capacity;

    public Queue(int capacity){
        arr = new int[capacity];
        this.capacity = capacity;
        inicio = 0;
        fim = 0;
        size = 0;
    }

    void enqueue(int val){
        if (size == capacity) {
            // Cria um novo array com o dobro do tamanho
            int[] newQueue = new int[capacity * 2];
            // Move os elementos para o novo array
            for (int i = 0; i < size; i++) {
                newQueue[i] = arr[(inicio + i) % capacity];
            }
            arr = newQueue;
            inicio = 0; // O início agora será 0, pois o array foi refeito
            fim = size; // O fim será igual ao tamanho da fila após a cópia
            capacity *= 2; // Atualiza a capacidade
        }
        arr[fim] = val;
        fim = (fim + 1) % capacity;
        size++;
    }

    void dequeue(){
        if (isEmpty()) {
            System.out.println("A fila está vazia.");
            return;
        }
        inicio = (inicio + 1) % capacity;
        size--;
    }

    boolean isEmpty(){
        return size == 0;
    }

    int peek(){
        if (isEmpty()) {
            System.out.println("A fila está vazia.");
            return -1; // Retorna -1 em caso de erro
        }
        return arr[inicio];
    }
}

File: DataStructures/Queue/Queue.rb
class Queue
  def initialize(capacity)
    @arr = Array.new(capacity)
    @capacity = capacity
    @inicio = 0
    @fim = 0
    @size = 0
  end

  def enqueue(val)
    if @size == @capacity
      new_arr = Array.new(@capacity * 2)
      @size.times do |i|
        new_arr[i] = @arr[(@inicio + i) % @capacity]
      end
      @arr = new_arr
      @inicio = 0
      @fim = @size
      @capacity *= 2
    end

    @arr[@fim] = val
    @fim = (@fim + 1) % @capacity
    @size += 1
  end

  def dequeue
    if is_empty?
      puts "A fila está vazia."
      return
    end
    @inicio = (@inicio + 1) % @capacity
    @size -= 1
  end

  def is_empty?
    @size == 0
  end

  def peek
    return nil if is_empty?

    @arr[@inicio]
  end
end
